File: partition-array-for-maximum-sum.js
// Time Complexity : O(A*K)
// Space Complexity : O(A)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : DP IS HARD

/**
 * @param {number[]} A
 * @param {number} K
 * @return {number}
 */
var maxSumAfterPartitioning = function(A, K) {
    if (!A || !A.length) return 0;
    if (!K) return 0;
    
    const N = A.length;
    const gain = new Array(N).fill(0);

    for (let i = 0; i < N; i++) {
        let maxElement = 0;
        for (let j = 1; j <= K; j++) {
            if (i - j + 1 >= 0) {
                maxElement = Math.max(maxElement, A[i - j + 1]);
                gain[i] = Math.max(gain[i], j * maxElement + (i - j >= 0 ? gain[i - j] : 0));
            }
        }
    }
    return gain[N - 1];
};

File: Partition_Array_for_Maximum_Sum.cpp
// Time Complexity : O(N*K) where N=Number of elements in array
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
//Dynamic Programming approach

class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& A, int K) {
        
        if(A.size()==0){
            return 0;
        }
        int m=A.size();
        vector<int> dp(m,0);
        
        for(int i=0;i<m;i++){
            int mx = A[i];
            for(int j=1;i-j+1>=0 && j<=K;j++){
                mx = max(mx,A[i-j+1]);
                if(i>=j)
                    dp[i] = max(dp[i],dp[i-j]+mx*j);
                else{
                    dp[i] = max(dp[i],mx*j);
                }
            }    
        }
        return dp[m-1];
    }
};
